[ONNX] Support optional type

garymm opened the pull request:
Based on #61938

neginraoof committed:
Add opset 14

neginraoof committed:
Update onnx build

neginraoof committed:
Fix for controlflow

neginraoof committed:
Fix for dynamic shape inference

neginraoof committed:
Fix controlflow

neginraoof committed:
Fix for controlflow

neginraoof committed:
Update opset 15

neginraoof committed:
Add _is_ op

neginraoof committed:
Merge

neginraoof committed:
Fix for comments

neginraoof committed:
Initial work for Optional type export

neginraoof committed:
Merge

neginraoof committed:
merge

neginraoof committed:
fix merge

neginraoof committed:
format

BowenBao committed:
[ONNX] Export nn.Module call as ONNX local function (#63589)

* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

ghstack-source-id: 92a95520cee1007c2b6d9c87698aba11b25147a1
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66140

neginraoof committed:
Merge commit 'refs/pull/61938/head' of https://github.com/pytorch/pytorch into neraoof/fixInputSequence

BowenBao committed:
[ONNX] Update onnx function export with comments and clean up (#66817)

* Addresses comments from #63589

Co-authored-by: BowenBao <bowbao@microsoft.com>

neginraoof committed:
Merge

BowenBao committed:
[ONNX] Update onnx function export with comments and clean up (#66817)

* Addresses comments from #63589

Co-authored-by: BowenBao <bowbao@microsoft.com>

[ONNX] remove torch::onnx::PRODUCER_VERSION (#67107)

Use constants from version.h instead.
This simplifies things since we no longer have to update
PRODUCER_VERSION for each release.

Also add TORCH_VERSION to version.h so that a string is available for
this purpose.

[ONNX] Set `ir_version` based on opset_version. (#67128)

This increases the odds that the exported ONNX model will be usable.
Before this change, we were setting the IR version to a value which may
be higher than what the model consumer supports.

Also some minor clean-up in the test code:
* Fix string replacement.
* Use a temporary file so as to not leave files around in the test
  current working directory.

Co-authored-by: Gary Miguel <garymiguel@microsoft.com>

ghstack-source-id: 2eb730ed6101bbf18b79b0941af4dbe8fcee1137
Pull Request resolved: https://github.com/pytorch/pytorch/pull/67803

garymm committed:
Undo ir_version changes

garymm committed:
Move unchecked cast into Prim class to match new style of namespacing

garymm committed:
restore third_party/onnx submodule to be in sync with onnx_ms_1

garymm committed:
minor cleanup

garymm committed:
undo bad merge

garymm committed:
Make skipScriptTest take a single opset version instead of list

garymm committed:
cleanup: Single quotes, comments

garymm committed:
enable tests. They are not passing though

garymm committed:
Handle ScriptFunction in _resolve_and_flatten_graph_inputs

fixes:
* TestExportModes.test_onnx_multiple_return
* TestONNXExport.test_export_tensoroption_to

Share the code with _decide_input_format

garymm committed:
minor cleanups

* rename input -> inputs for clarity
* remove redundant test
* fix comment about why script test is disabled

garymm committed:
minor cleanup

* use list.append instead of +=
* tighten try / except block

garymm committed:
Fix TestONNXRuntime_opset15::test_optional_inputs_with_no_optionals

Handle empty kwargs for script modules

garymm committed:
rename and simplify test

garymm committed:
fix tests

garymm committed:
fix tests with empty inputs. Use isinstance for checking against NoneType and OptionalType

garymm commented:
@BowenBao can you review since you already started reviewing https://github.com/pytorch/pytorch/pull/61938?
If you're busy let me know and I'll ask someone else.

BowenBao commented on torch/csrc/jit/serialization/export.cpp:
```diff
@@ -1083,6 +1134,41 @@ void GraphEncoder::EncodeLocalFunction(
+      onnx::TensorShapeProto* shape = tensor_type_proto->mutable_shape();
+      auto sizes = t->symbolic_sizes().sizes().value();
+      for (const auto i : c10::irange(sizes.size())) {
+        shape->add_dim();
```
nit: a little duplicated code with `EncodeValueInfoType` that can be extracted.

BowenBao commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -3398,7 +3398,8 @@ def Loop(ctx: torch.onnx.SymbolicContext, g, *inputs, **attrs):
                     b_in.setType(inputs[i].type())
-                if i > 0 and (i + 1) < len(inputs):
+                # For optional block inputs, input type could be NoneType, which we don't want to propagate.
+                if i > 0 and (i + 1) < len(inputs) and not isinstance(inputs[i + 1].type(), NoneType):
```
what happens for using optional block inputs with loop opset < 15. This code looks like it will be silently not setting types for it. Will this end up with invalid model?

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -525,6 +547,9 @@ def _model_to_graph(model, args, verbose=False,
 
     _set_input_and_output_names(graph, input_names, output_names)
     params_dict = _get_named_param_dict(graph, params)
+    # make sure that the param dict and the graph match each other
```
Do you know what's the rationale behind the changes around this block?

BowenBao reviewed (commented):
Thank you @garymm for updating this pull request! I'm leaving some comments and thoughts as looking through the code.

garymm committed:
[ONNX] Minor doc update (#69501)

Fix the wiki URL.

Also minor reorganization in onnx.rst.

BowenBao committed:
[ONNX] De-duplicate initializers (#68202)

ScriptModule export introduces duplicated ONNX initializers for shared weights, unnecessarily increases ONNX model size. This PR de-duplicates ONNX initializers for model exported in eval mode, by checking if the underlying tensors share the same `data_ptr`, `strides` and `sizes`.

Co-authored-by: BowenBao <bowbaomicrosoft.com>

ghstack-source-id: d17dfa4adbe65bf410d092267676753555656d57
Pull Request resolved: https://github.com/pytorch/pytorch/pull/69547

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -970,77 +958,108 @@ def forward(self, x, y=None, z=None):
+            self.run_test(model, (x, {"y": y}), input_names=("x", "y"))
+
+        # Setting any input to None in tracing means that
+        # input doesn't exist at all in the exported model, so input_names doesn't match.
```
todo: check what is the current warning/error message.

BowenBao replied:
current error message: "RuntimeError: number of input names provided (3) exceeded number of inputs (2)"

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -190,14 +190,26 @@ std::vector<Value*> ConvertSequenceDependencies(Node* node, int opset_version) {
-        block->return_node()->replaceInputWith(output, id_node->output());
+        if (auto none_output_type = output->type()->cast<NoneType>() &&
+                index + 1 < block->inputs().size()) {
+          auto t = block->inputs().at(index + 1)->type()->cast<OptionalType>();
```
This seems to be covering cases where identity takes optional input, which should be covered by identity in opset 16.
todo:  try remove and evaluate if existing test case failed. To keep this at least a fix to increment index through looping is needed.

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -235,6 +247,20 @@ void FixupONNXLoopNodeInputs(Node* node) {
   }
+
+  size_t index = 0;
+  for (auto input : node->inputs()) {
```
note: reactively convert aten None to onnx Optional, when they are inputs to loop.

garymm committed:
[ONNX] restore documentation of public functions (#69623)

The build-docs check requires all public functions to be documented.
These should really not be public, but we'll fix that later.'

shubhambhokare1 committed:
[ONNX] Add module name as PythonOp attribute (#67193)

* Add module name as pythonOp attr

* Move to trace_post_record

* Add tests

* Code compactness

garymm committed:
Got test_loop_none_output to seemingly produce a valid graph, though
ORT doesn't support op set 16 yet so it still fails at run time.

garymm committed:
Simplify test cases and refactor and comment C++ passes

Lots of debugging code still in at the moment.

garymm committed:
Lots of fixes

TODO next:
* Improve test_loop_none_output
* Remove debug prints

garymm committed:
I think everything is working.

* Fix some more control flow cases.
* Move tests to a new file.
* Remove debug cerr output.

garymm committed:
comments

garymm committed:
parameterize test case and add cases for condition is false

garymm committed:
nest parameterize

garymm committed:
cleanup and fix a bug with input vs inputs

garymm committed:
comment explaining whats going on, but havent fixed it yet

garymm committed:
Pip through is_script into _jit_pass_onnx_assign_output_shape

Currently failing:
python test/onnx/test_pytorch_onnx_onnxruntime.py TestONNXRuntime.test_tuple_with_none_outputs

Fails with:
torch.onnx.CheckerError: Unrecognized type value case (value_info name: 1): 0

garymm committed:
Fix run_ort, lots of other clean up.

Nothing I know that is still broken or missing.

garymm committed:
mostly fix rebase bugs

garymm reviewed (commented):
@BowenBao I think this is ready for review. The check failures seem unrelated.

LMK if you'd like to do this in real time together.

BowenBao commented:
Thanks @garymm, I will try to read through the code changes first.

BowenBao commented:
Conflicting files due to out of sync, @garymm could you please do another rebase with latest onnx_ms_1 branch?

garymm commented:
@BowenBao rebased

BowenBao commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,80 @@
+        self.assertEqual(len(output_0_tensor_type.shape.dim), len(x.shape))
+
+
+instantiate_parametrized_tests(TestOptionalOutput)
```
Thanks for creating this test file. The goal of testing export without running exported model in backend seems to align with many test cases in test_utility_funs.py, and those test cases are more suited in this file. Should we move them over in next steps?

garymm replied:
Sure. I think we should also move over tests from test/jit/test_onnx_export.py. Tracked by https://msdata.visualstudio.com/Vienna/_workitems/edit/1593350.

BowenBao commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,80 @@
+        output_0_optional_type = output_0_type.optional_type
+        self.assertTrue(output_0_optional_type.elem_type.HasField("tensor_type"))
+        output_0_tensor_type = output_0_optional_type.elem_type.tensor_type
+        self.assertEqual(
```
This part seems very verbose. Can it be done by direct onnx TypeProto comparison, with hope that onnx TypeProto construction of optional_type with the expected element tensor type is straightforward and simple?

BowenBao commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,80 @@
+            output_0_tensor_type.elem_type,
+            symbolic_helper.scalar_type_to_onnx[
+                symbolic_helper.scalar_type_to_pytorch_type.index(x.dtype)])
+        self.assertEqual(len(output_0_tensor_type.shape.dim), len(x.shape))
```
Should test cover what dimension size is? is it a symbol or empty?

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -109,17 +108,30 @@ def inline_flatten_list(inputs, res_list):
+        if input_name in ort_inputs:
+            raise ValueError(
+                f"got too many positional inputs. inputs: {inputs}. kw_inputs: {kw_inputs}")
+        ort_inputs[input_name] = input
```
are the above change mainly for covering extra dict input that weren't covered in previous prs, and unrelated to optional input?

edit: looking at other parts of code, looks like this is required for optional.

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -824,23 +839,23 @@ def forward(self, x_in):
     def test_dict_str(self):
         class MyModel(torch.nn.Module):
-            def forward(self, x_in):
+            def forward(self, x_in: Dict[str, Tensor]) -> Dict[str, Tensor]:
```
does this change affect exporting?

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -978,88 +963,124 @@ def forward(self, x, y=None, z=None):
-    def test_input_as_output(self):
+        x = torch.randn(2, 3)
+        model = Model()
+        self.run_test(model, (x, None))
```
With tracing, do these tests capture `if` operator in exported graph?

I wonder if it'd work when providing extra set of inputs (e.g. (None, other_tensor)) to `run_test`.

BowenBao replied:
from comments looks like the answer to above question is it won't capture if and it won't work with other inputs combination in tracing ... 

Is it reasonable to assume that for optional input feature to really be effective, it would require scripting?

BowenBao replied:
Thanks for explanation. Let's keep it to ensure tracing route is compatible with this signature and None/Optional usage.

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -978,88 +963,124 @@ def forward(self, x, y=None, z=None):
 
         x = 3
-        y = torch.randint(10, (2, 3, 4))
+        y = torch.ones(2, 3, 4)
```
any particular reason for `randint -> ones`?
edit: dtype of y becomes float instead of int with this change.

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -1732,12 +1753,13 @@ def forward(self, x: int, y: int):
         self.run_test(ArithmeticModule(), (x, y))
 
-    @disableScriptTest()
+    # In tracing, None outputs are removed. In scripting they're kept but
```
does this mean we can enable test for optional outputs? e.g. None in one branch, some tensor in other.

garymm replied:
Yes. Tests for that are in test/onnx/test_pytorch_onnx_no_runtime.py

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -8327,7 +8340,7 @@ def make_input(batch_size):
         self.run_test(model, other_input, batch_size=RNN_BATCH_SIZE + 1)
 
-    @disableScriptTest()  # TODO: RuntimeError: Exporting the operator __is_ to ONNX is not supported
+    @skipScriptTest()  # TODO: https://msdata.visualstudio.com/Vienna/_workitems/edit/1253950
```
__is_ operator is already added. Is this still failing but for different reason?

garymm replied:
Yes, currently fails with:
```
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                                │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 8360, in test_transformer_encoder                                                                                                                                                                                                                                   │·········
    self.run_test(MyModule(512, 8, 2048 , 0., 3), (x,), atol=1e-6)                                                                                                                                                                                                                                                                                                │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 347, in run_test                                                                                                                                                                                                                                                    │·········
    _run_test(script_model, scripting_remained_onnx_input_idx, flatten=False)                                                                                                                                                                                                                                                                                     │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 327, in _run_test                                                                                                                                                                                                                                                   │·········
    return run_model_test(self, m, batch_size=batch_size,                                                                                                                                                                                                                                                                                                         │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 185, in run_model_test                                                                                                                                                                                                                                              │·········
    ort_sess = convert_to_onnx(model, input=input, opset_version=self.opset_version,                                                                                                                                                                                                                                                                              │·········
  File "/home/azureuser/src/pytorch/pytorch/test/onnx/test_pytorch_onnx_onnxruntime.py", line 88, in convert_to_onnx                                                                                                                                                                                                                                              │·········
    torch.onnx._export(model, input_copy, f,                                                                                                                                                                                                                                                                                                                      │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/__init__.py", line 50, in _export                                                                                                                                                                                                                                                                          │·········
    result = utils._export(*args, **kwargs)                                                                                                                                                                                                                                                                                                                       │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 751, in _export                                                                                                                                                                                                                                                                            │·········
    _model_to_graph(model, args, verbose, input_names,                                                                                                                                                                                                                                                                                                            │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 540, in _model_to_graph                                                                                                                                                                                                                                                                    │·········
    graph = _optimize_graph(graph, operator_export_type,                                                                                                                                                                                                                                                                                                          │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 229, in _optimize_graph                                                                                                                                                                                                                                                                    │·········
    graph = torch._C._jit_pass_onnx(graph, operator_export_type)                                                                                                                                                                                                                                                                                                  │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/__init__.py", line 368, in _run_symbolic_function                                                                                                                                                                                                                                                          │·········
    return utils._run_symbolic_function(*args, **kwargs)                                                                                                                                                                                                                                                                                                          │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/utils.py", line 1104, in _run_symbolic_function                                                                                                                                                                                                                                                            │·········
    return symbolic_fn(g, *inputs, **attrs)                                                                                                                                                                                                                                                                                                                       │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/symbolic_opset10.py", line 27, in div                                                                                                                                                                                                                                                                      │·········
    return torch.onnx.symbolic_opset9.true_divide(g, self, other)                                                                                                                                                                                                                                                                                                 │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/symbolic_opset9.py", line 196, in true_divide                                                                                                                                                                                                                                                              │·········
    if sym_help._is_fp(self) or sym_help._is_fp(other):                                                                                                                                                                                                                                                                                                           │·········
  File "/home/azureuser/src/pytorch/pytorch/torch/onnx/symbolic_helper.py", line 340, in _is_fp                                                                                                                                                                                                                                                                   │·········
    type = value.type().scalarType()                                                                                                                                                                                                                                                                                                                              │·········
RuntimeError: rINTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h":540, please report a bug to PyTorch.
```

BowenBao reviewed (commented):
Thank you for taking over this PR. Still reading through changes, leaving questions/comments along the way

garymm committed:
clean up

garymm reviewed (commented):
Thanks for the review.

garymm commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -504,9 +670,14 @@ std::vector<Value*> FixupONNXControlflowNode(Node* n, int opset_version) {
+        // to handle the case where a block input is Optional but the
+        // output is not (i.e. if the loop executes > 0 times, the
+        // output will not be None).
+        auto type = n->blocks().at(0)->inputs().at(i + 2)->type();
```
Tests are failing at this line in the case where `ConvertSequenceDependencies` is triggered because that breaks the expectation that the block inputs contain all the block outputs. I'm looking into how to fix this. Let me know if you have ideas.

garymm replied:
Fixed!

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -321,17 +333,20 @@ def _run_test(m, remained_onnx_input_idx, flatten=True):
-            script_model = torch.jit.script(model)
+        is_script = isinstance(model, (torch.jit.ScriptModule, torch.jit.ScriptFunction))
+
+        if self.is_script_test_enabled:
```
TODO: audit tests that have `@skipScriptTest` but use `torch.jit.script` or `torch.jit.trace` explicitly in test body.

garymm committed:
handle scan outputs in FixupONNXControlflowNodeOutputs

garymm reviewed (commented):
@BowenBao I fixed the failing tests. Hopefully everything is good to go now. Please take a look. I'd really like to get this merged next week if possible.

garymm committed:
fix run_ort arg name

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -401,11 +405,34 @@ def _get_param_count_list(method_graph, args_params):
+    resolved_args = []  # type: ignore[var-annotated]
+    if isinstance(model, (torch.jit.ScriptFunction, torch.jit.ScriptModule)):
+        for i, var in enumerate(args_params):
+            if var is None:
```
I feel the purpose is to grab default value when argument `var` is omitted in example inputs. But how does it work? I thought `var is None` is supposed to mean `None` passed as input explicitly. Let me know where I'm wrong.

BowenBao commented on torch/onnx/utils.py:
```diff
@@ -401,11 +405,34 @@ def _get_param_count_list(method_graph, args_params):
+        for i, var in enumerate(args_params):
+            if var is None:
+                resolved_args.append(sig.parameters[param_keys[i]].default)
+            elif isinstance(var, tuple):
```
nit: not sure if nested tuple is something we'd care about.

garymm replied:
Moot now that this function is gone.

garymm committed:
Don't allow None example_inputs

garymm reviewed (commented):
Thanks for the review!

garymm committed:
skip self in signature and dont append args when no default value provided

BowenBao commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,76 @@
+# Owner(s): ["module: onnx"]
```
Add to `scripts/onnx/test.sh` if intend to run in CI.

garymm commented:
Converting to draft until I have time to understand the shape and type inference issues better and add more tests.

garymm committed:
fix dropout_training tests

garymm committed:
Fix and test for mismatched types between if branches

garymm commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -1828,6 +1823,14 @@ void SpecialPostProcess(Node* n) {
       }
       break;
     }
+    case ::c10::onnx::Optional: {
```
Deleted. It seems it may not be needed.

garymm commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -237,6 +241,12 @@ bool IsValidONNXNode(const Node* n) {
 
+  for (auto inp : n->inputs()) {
+    if (inp->type() == NoneType::get()) {
+      return false;
```
Found and fixed the only known case in `FixupONNXSubblockOutputs` by using empty `onnx::Optional` instead of `onnx::Identity(None)`. Will address other cases as needed.

garymm committed:
test for uninitialized optional output

garymm committed:
fix _decide_input_format when input is tensor

garymm committed:
remove unneeded shape type inference code

garymm reviewed (commented):
If CI passes I think this might be ready to go!

garymm committed:
lint

garymm committed:
add type annotation for args_dict

BowenBao reviewed (commented):
Some test failures from `test_pytorch_onnx_caffe2.py`. It looks having to do with how input arguments are parsed.

garymm committed:
Fix None arg checking: don't check params, just args

BowenBao reviewed (approved):
Thanks again for completing this PR!

garymm committed:
handle list of args, work around mypy

garymm committed:
fix Caffe2 LSTM input args

garymm committed:
fix error message in test. remove unneded check for NoneType in symbolic

garymm committed:
expand comment

garymm committed:
use block input type only when needed

garymm committed:
always use block output type, wrapped in optional if needed

garymm reviewed (commented):
Fixed the last remaining issues. Going to merge now.
